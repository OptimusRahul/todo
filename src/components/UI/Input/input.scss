.Input {
    display: flex;
    justify-content: center;

    width: 100%;
    height: 2rem;

    margin: 2rem 0;

    &--box {
        width: 75%;
        height: 2.5rem;
        border: none;
        outline: none;
        
        border-bottom: 3px solid #797876;

        border-radius: 5px;
        background: #ffffff;
        box-shadow:  5px 5px 10px #d4d4d4, -5px -5px 10px #ffffff;

        font-family: "Lato", sans-serif;

        &:active, &:focus {
            border-bottom: 3px solid #1244a0;
        }
    }
}