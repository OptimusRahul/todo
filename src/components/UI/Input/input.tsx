import { ChangeEvent, FC, useRef, useState } from 'react';
import { useContainer } from 'unstated-next';

import { generateTodoObject } from '../../../utils';
import TodoContext from '../../../contexts/todo.context';
import './input.scss';

const Input: FC = (): JSX.Element => {
    const inputRef = useRef<HTMLInputElement>(null);
    const [ newTask, setNewTask ] = useState<string>();

    const { setTodo } = useContainer(TodoContext);

    /**
     * 
     * @param e : ChangeEvent<HTMLInputElement>
     * 
     * @description: Take Input box value and update it to the memory
     * 
     */

    const onInputChangeHandler: Function = (e: ChangeEvent<HTMLInputElement>): void => setNewTask(e.target.value);

    /**
     * 
     * @param e : KeyboardEvent
     * 
     * @description: Check for 'Enter' key to update the new task
     * 
     */
    
    const addNewTodoHandler: Function = (e: KeyboardEvent): void => {
        if(e.key === 'Enter') {
            if(newTask && newTask.trim()) {
                const todo = generateTodoObject(newTask);
                setTodo(todo);
                setNewTask('');
            } else {
                window.alert('Please enter task');
            }
            if(inputRef.current) {
                inputRef.current.value = '';
            }
        }
    }

    return (
        <div className="Input" data-testid="Input--test">
            <input
                ref={inputRef}
                className="Input--box" 
                data-testid="Input--box__test"
                placeholder="Please enter todo here" 
                width="400"
                height="100"
                onKeyDown={(e) => addNewTodoHandler(e)}
                onChange={(e) => onInputChangeHandler(e)}
                defaultValue={""}
                />
        </div>
    )
}

export default Input;