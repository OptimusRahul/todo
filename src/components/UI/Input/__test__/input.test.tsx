import React from 'react'; 
import renderer from 'react-test-renderer';
import { fireEvent, render, cleanup } from '@testing-library/react';

import Input from '../input'; 
import TodoContext from '../../../../contexts/todo.context';

afterEach(cleanup);

test('renders correctly', () => {
    const { queryByTestId, queryByPlaceholderText } = render(
        <TodoContext.Provider>
            <Input />
        </TodoContext.Provider>
    );
    expect((queryByTestId("Input--test"))).toBeInTheDocument();
    expect(queryByPlaceholderText('Please enter todo here')).toBeTruthy();

});

describe('on input change', () => {
    it('changes value correctly', () => {
        const { getByTestId } = render(
            <TodoContext.Provider>
                <Input />
            </TodoContext.Provider>
        )

        const todoInput = getByTestId("Input--box__test");
        
        fireEvent.change(todoInput, { target: { value : 'test' } });

        expect(todoInput).toHaveValue('test');
    });

    it('matches changed value to the state', () => {
        const tree = renderer.create(
            <TodoContext.Provider>
                <Input />
            </TodoContext.Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});