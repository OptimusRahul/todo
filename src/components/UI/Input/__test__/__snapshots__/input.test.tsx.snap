// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on input change matches changed value to the state 1`] = `
<div
  className="Input"
  data-testid="Input--test"
>
  <input
    className="Input--box"
    data-testid="Input--box__test"
    defaultValue=""
    height="100"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Please enter todo here"
    width="400"
  />
</div>
`;
